### PR TITLE
Caroline Nardi - API Muncher - Octos

### PR DESCRIPTION
**Pull request by cmn53**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I looked through the documentation and then sent a general query request and a specific recipe request in Postman to explore the response structure.
Describe your API Wrapper.  How did you decide on the methods you created? | My API wrapper has two methods, one for calling the Edamam API with a general query and returning a collection of recipes, and one for requesting a single recipe from the API and returning an instance of class Recipe. I decided on two methods because I felt the parameters for each API call differed enough to necessitate distinct methods.
Describe an edge case or failure case test you wrote for your API Wrapper. | I wrote edge case tests for invalid or absent queries or recipe ids (although I wasn't successful in parsing out the API's response codes, so these all display the same generic error message whether the call fails or there just aren't any results from the query.
Explain how VCR aids in testing an API. | VCR records HTTP responses so that the tests do not have to make calls directly to the API. This helps you to not rack up unnecessary API calls during testing or have failing tests if the API is down.
What is the Heroku URL of your deployed application? | https://adie-wok-in-progress.herokuapp.com/
Provide a link to the Trello board you used | https://trello.com/b/RrI9Pe3V/api-muncher-public

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  root 'recipes#search'
+  get '/results', to: 'recipes#index', as: 'results'
+  get '/recipes/:id', to: 'recipes#show', as: 'recipe'
```
Comment (droberts-sea) [inline config/routes.rb:5]: Could you use `resources` for this?

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  def show
+    id = params[:id]
+    @recipe = EdamamApiWrapper.recipe_details(id)
+  end
+
+  private
+  def catch_api_error
+    begin
+      yield
+    rescue EdamamApiWrapper::EdamamError
```
Comment (droberts-sea) [inline app/controllers/recipes_controller.rb:29]: Good work matching your error handling here to what's going on in the API wrapper.

File: lib/edamam_api_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
+require 'httparty'
+
+class EdamamApiWrapper
+  class EdamamError < StandardError; end
+
+  APP_ID = ENV["EDAMAM_APP_ID"]
+  APP_KEY = ENV["EDAMAM_APP_KEY"]
+  BASE_URL = "https://api.edamam.com/search?"
+  BASE_URI = "http://www.edamam.com/ontologies/edamam.owl#recipe_"
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:9]: The variable names `BASE_URL` and `BASE_URI` are very similar - I would be worried about typoing one for the other. Could you choose something more visually distinct (maybe `BASE_RECIPE_ID`)?

File: lib/edamam_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  def self.search_results(query)
+    url = "#{BASE_URL}q=#{query}&app_id=#{APP_ID}&app_key=#{APP_KEY}&from=0&to=120"
+
+    response = HTTParty.get(url)
+
+    unless response["count"] > 0
+      raise EdamamError.new
+    end
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:18]: You should include an error message here, so that if some other developer (or future you) encounters this error they know what went wrong.

File: test/controllers/recipes_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      get root_path
+
+      must_respond_with :success
+    end
+  end
+
+  describe "index" do
+    it "succeeds for a valid search query" do
+      VCR.use_cassette('recipes') do
+        get results_path, params: { query: "chicken"}
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:15]: It might also be worthwhile to add some tests around the paging parameters:

- What happens if they're absent?
- Do you get different results when they change?
- What if you send a bogus value, like a negative page number?

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | yes
Comprehension questions | yes
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | yes
Errors are reported to the user | yes
API Wrapper to handle the API requests | yes
Controller testing | yes
Lib testing | yes
Search Functionality | yes
List Functionality | yes
Show individual item functionality (link to original recipe opens in new tab) | yes
**Styling** | 
Responsive layout | yes
List View shows 10 items at a time/pagination | yes
The app is styled to create an attractive user interface | yes
**API Features** | 
The App attributes Edaman | yes
The VCR casettes do not contain the API key | yes
**External Resources** | 
Link to Trello Board | yes
Link to deployed app on Heroku | yes
**Overall** | Excellent job overall! This code is well-organized and easy to read, and it's clear the learning goals for this assignment were met. I've left some nitpicks inline below, but in general I'm quite happy with this submission. Keep up the hard work.

**Comment by droberts-sea** [comment]
Also, I love your site name.
